tsl/src/nodes/decompress_chunk/planner.c

```diff
@@ -40,169 +41,219 @@ _decompress_chunk_init(void)
+		chunk_attrs_found =
+			bms_add_member(chunk_attrs_found,
+						   TableOidAttributeNumber - FirstLowInvalidHeapAttributeNumber);
 	}
-	else
-		scan_var = makeVar(path->info->compressed_rel->relid,
-						   scan_varattno,
-						   ts_custom_type_cache_get(CUSTOM_TYPE_COMPRESSED_DATA)->type_oid,
-						   -1,
-						   0,
-						   0);
-	path->decompression_map = lappend_int(path->decompression_map, chunk_attno);
-
-	return makeTargetEntry((Expr *) scan_var, tle_index, NULL, false);
-}
-
-/*
- * build targetlist for scan on compressed chunk
- *
- * Since we do not adjust selectedCols in RangeTblEntry for chunks
- * we use selectedCols from the hypertable RangeTblEntry to
- * build the target list for the compressed chunk and adjust
- * attno accordingly
- */
-static List *
-build_scan_tlist(DecompressChunkPath *path)
-{
-	List *scan_tlist = NIL;
-	Bitmapset *attrs_used = path->info->ht_rte->selectedCols;
-	TargetEntry *tle;
-	int bit;
 
-	path->decompression_map = NIL;
+	/*
+	 * Fill the helper array of compressed attno -> compression info.
+	 */
+	FormData_hypertable_compression **compressed_attno_to_compression_info =
+		palloc0(sizeof(void *) * (path->info->compressed_rel->max_attr + 1));
+	ListCell *lc;
+	foreach (lc, path->info->hypertable_compression_info)
+	{
+		FormData_hypertable_compression *fd = lfirst(lc);
+		AttrNumber compressed_attno =
+			get_attnum(path->info->compressed_rte->relid, NameStr(fd->attname));
 
-	/* add count column */
-	tle = make_compressed_scan_meta_targetentry(path,
-												COMPRESSION_COLUMN_METADATA_COUNT_NAME,
-												DECOMPRESS_CHUNK_COUNT_ID,
-												list_length(scan_tlist) + 1);
-	scan_tlist = lappend(scan_tlist, tle);
+		if (compressed_attno == InvalidAttrNumber)
+		{
+			elog(ERROR,
+				 "column '%s' not found in the compressed chunk '%s'",
+				 NameStr(fd->attname),
+				 get_rel_name(path->info->compressed_rte->relid));
+		}
 
-	/* add sequence num column */
-	if (path->needs_sequence_num)
-	{
-		tle = make_compressed_scan_meta_targetentry(path,
-													COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME,
-													DECOMPRESS_CHUNK_SEQUENCE_NUM_ID,
-													list_length(scan_tlist) + 1);
-		scan_tlist = lappend(scan_tlist, tle);
+		compressed_attno_to_compression_info[compressed_attno] = fd;
 	}
 
-	/* check for system columns */
-	bit = bms_next_member(attrs_used, -1);
-	if (bit > 0 && bit + FirstLowInvalidHeapAttributeNumber < 0)
+	/*
+	 * Go over the scan targetlist and determine to which output column each
+	 * scan column goes.
+	 */
+	path->decompression_map = NIL;
+	foreach (lc, scan_tlist)
 	{
-		/* we support tableoid so skip that */
-		if (bit == TableOidAttributeNumber - FirstLowInvalidHeapAttributeNumber)
-			bit = bms_next_member(attrs_used, bit);
+		TargetEntry *target = (TargetEntry *) lfirst(lc);
+		if (!IsA(target->expr, Var))
+		{
+			elog(ERROR, "compressed scan targetlist entries must be Vars");
+		}
```

Comment:
Can you create a test case for this? Not sure if it is possible.

Reply:
No, this would be a logic error.

Reply:
Ok